Enonic UI: UnpublishDialog #9532

skjulestad opened the pull request:
Migratet UnpublishDialog, moved description back to header. Changed order of footer buttons to fit figma design.

skjulestad committed:
Enonic UI: UnpublishDialog #9532

Migratet UnpublishDialog, moved description back to header.
Changed order of footer buttons to fit figma design.

edloidas commented on modules/app/src/main/resources/assets/js/main.ts:
```diff
@@ -374,17 +375,11 @@ async function startApplication() {
+        openUnpublishDialog(event.getModels(), {
+            onYes: event.getYesCallback(),
+            onNo: event.getNoCallback(),
+        });
```
We no longer use `ContentUnpublishPromptEvent` to control the dialog state.
Remove it, it's handlers, and replace it's usage in Actions with direct `openUnpublishDialog()` call

edloidas commented on modules/lib/src/main/resources/assets/js/app/browse/ContentUnpublishPromptEvent.ts:
```diff
@@ -5,10 +5,32 @@ import {ContentSummaryAndCompareStatus} from '../content/ContentSummaryAndCompar
 
 export class ContentUnpublishPromptEvent extends BaseContentModelEvent {
```
Remove completelly

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/shared/dialogs/delete/DeleteDialog.tsx:
```diff
@@ -112,15 +112,22 @@ export const DeleteDialog = (): ReactElement => {
             <>
-                <Dialog.DefaultHeader title={confirmTitle} withClose />
+                <Dialog.DefaultHeader title={confirmTitle}>
+                    <p className="text-subtle">{confirmDescription}</p>
```
Shouldn't it be just in `description` in DefaultHeader? Also, text is not subtle by design.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/shared/dialogs/unpublish/UnpublishDialog.tsx:
```diff
@@ -0,0 +1,205 @@
+                                                content={item}
+                                                branch={Branch.DRAFT}
+                                                hasInbound={isInbound(item)}
+                                            />
```
Saem here

skjulestad replied:
This got refactored to reflect DeleteDialog.tsx, I do not understand what change is wanted here.

skjulestad replied:
Patched in next commit

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/shared/dialogs/unpublish/UnpublishDialog.tsx:
```diff
@@ -0,0 +1,205 @@
+    };
+
+    return (
+        <>
```
Why do you need to wrap it in Fragment? You return just a single component.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/shared/dialogs/unpublish/UnpublishDialog.tsx:
```diff
@@ -0,0 +1,205 @@
+                                        onIgnore: () => ignoreUnpublishInboundDependencies(),
+                                    },
+                                }}
+                            />
```
I see we use Status such way twice already. Maybe it's better to adjust SelectionStatusBar or create a new component, so that we don't need to initialize it with lots of "nullish" data.

skjulestad replied:
This will be refactored in next commit as well, but after removing showReferences I am starting to wonder if we even need SelectionStatusBar for UnpublishDialog, at least we have it and can remove it if we want then

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/shared/dialogs/unpublish/UnpublishDialog.tsx:
```diff
@@ -0,0 +1,205 @@
+                            <Dialog.Body className="flex flex-col gap-y-10">
+                                <ul className="flex flex-col gap-y-2.5">
+                                    {items.map(item => (
+                                        <ContentListItemWithReference
```
I specifically mentioned that Unpublish dialog shows regular content without "Show Reference" links.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/shared/dialogs/unpublish/UnpublishDialog.tsx:
```diff
@@ -0,0 +1,205 @@
+            <Dialog.Root open={open} onOpenChange={handleOpenChange}>
+                <Dialog.Portal>
+                    <Dialog.Overlay/>
+                    {view === 'confirmation' ? (
```
You will have progress step in the future. Using ternary here is not a good idea, as you'll have to check for more than two views. Just check for it directly and render the corresponding body `view === 'confirmation' && <`

edloidas replied:
And it's better to place confirmation after main dialog.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+        $unpublishDialogPending.set(initialPendingState);
+        $unpublishDialog.set({
+            ...viewState,
+            open: true,
```
Don't open dialog on fail. We don't have multiple roots, we have multiple content components. It's either already open, or closed by the user.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+    inboundTargets: ContentId[];
+    inboundIgnored: boolean;
+    yesCallback?: () => void;
+    noCallback?: () => void;
```
Remove and recheck the logic in the store.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+import {$contentUnpublished} from '../socket.store';
+import {createDebounce} from '../../utils/timing/createDebounce';
+
+export type UnpublishAction = 'unpublish';
```
You have one action only. Remove it. We can simplify it and make it closer to publish.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+    return [...items, ...dependants].some(item => item.getContentSummary().isSite());
+});
+
+export const $unpublishHasBlockingInbound = computed($unpublishDialog, ({inboundTargets, inboundIgnored}) => {
```
Must have different names following best practices in naming and start with `$is*`. Check the Delete dialog store.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+
+export const $unpublishInboundIds = computed($unpublishDialog, ({inboundTargets}) => inboundTargets.map(id => id.toString()));
+
+export const $unpublishDialogReady = computed([$unpublishDialog, $unpublishHasBlockingInbound], (state, hasInbound) => {
```
This name too

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+// * Helpers
+//
+
+const sortDependants = (dependants: ContentSummaryAndCompareStatus[], inboundTargets: ContentId[]): ContentSummaryAndCompareStatus[] => {
```
Looks similar to Delete dialog. Probably can be extracted to utils.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+        return;
+    }
+
+    instanceId += 1;
```
Instance must be incremented not here, but when the new request starts.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+    $unpublishDialog.set({
+        ...initialViewState,
+        open: true,
+        loading: true,
```
Loading must be handled via subscription `$unpublishDialog.subscribe(({open}, oldState) => ...);`
See delete and unpublish dialogs.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+    instanceId += 1;
+
+    $unpublishDialog.set({
+        ...initialViewState,
```
See how it's done in Delete dialog store. You should use `structuredClonde` here, since `initialViewState` has mutable properties and applying it with spread operator you still keep the original arrays from the initial state, that can be accidentally mutated.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+    } else if (!open && wasOpen) {
+        instanceId += 1;
+    }
+});
```
Update it to work the same as in other stores.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+    }
+});
+
+$contentUnpublished.subscribe((event) => {
```
Still missing `$contentUpdated`, `$contentCreated`, `$contentArchived`, `$contentDeleted` handlers to update and reload the lists on server content changes.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+export const $unpublishInboundIds = computed($unpublishDialog, ({inboundTargets}) => inboundTargets.map(id => id.toString()));
+
+export const $unpublishDialogReady = computed([$unpublishDialog, $unpublishHasBlockingInbound], (state, hasInbound) => {
+    const {submitting} = $unpublishDialogPending.get();
```
Non-reactive. Computed stores must not get data from other stores via `get()`. Add store to the stores array.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+};
+
+// Initial state snapshot for reset
+const initialViewState: UnpublishDialogStore = {
```
Probably misleading name, as it's not view state. Just `initialState` or even `INITIAL_STATE` would be good.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/dialogs/unpublishDialog.store.ts:
```diff
@@ -0,0 +1,385 @@
+};
+
+export const executeUnpublishDialogAction = async (): Promise<boolean> => {
+    const viewState = $unpublishDialog.get();
```
Just get the data from store directly via destructing `{loading, failed, ...}`. No need to restore the state of the dialog on error. It'll either stay unchanged, or user will select another data and open unpublish dialog again just to see that after failed publishing the old content will be restored.

skjulestad committed:
Enonic UI: UnpublishDialog #9532

Migratet UnpublishDialog, moved description back to header.
Changed order of footer buttons to fit figma design.
Refactored after pr review.

skjulestad commented:
There is now an open pr for implementing progress bar based on this branch: https://github.com/enonic/app-contentstudio/pull/9546

edloidas reviewed (approved):
Good work. There are still some problems that need to be addressed, but they are common to several dialogs and rather big, so will be done in a separate task.
